service/history/historyEngine.go

```diff
@@ -1628,6 +1659,9 @@ Update_History_Loop:
 					return nil, &workflow.BadRequestError{Message: fmt.Sprintf("Unknown decision type: %v", *d.DecisionType)}
 				}
 			}
+			if err := sizeChecker.failWorkflowIfTotalSizeExceedsLimit(); err != nil {
```

Comment:
Looks like the existing logic already has some bugs.  We are not setting the isComplete flag which will result us not creating the appropriate timer and transfer tasks for completed workflows.